cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -76,11 +79,32 @@ struct IsInfOperator {
 
 struct IsNullOperator {
   static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
-    checked_cast<BooleanScalar*>(out)->value = !in.is_valid;
+    auto options = OptionsWrapper<NanNullOptions>::Get(ctx);
+    bool* out_value = &checked_cast<BooleanScalar*>(out)->value;
+    if (in.is_valid) {
+      switch (in.type->id()) {
+        case Type::FLOAT:
+          *out_value = options.nan_is_null &&
+                       std::isnan(internal::UnboxScalar<FloatType>::Unbox(in));
+          break;
+        case Type::DOUBLE:
+          *out_value = options.nan_is_null &&
+                       std::isnan(internal::UnboxScalar<DoubleType>::Unbox(in));
+          break;
+        default:
+          *out_value = false;
+      }
+    } else {
+      *out_value = true;
+    }
+
     return Status::OK();
   }
 
   static Status Call(KernelContext* ctx, const ArrayData& arr, ArrayData* out) {
+    // TODO: Is `options` needed for detect nulls? Which is the better way to
+    // handle is_null for ArrayData
+    auto options = OptionsWrapper<NanNullOptions>::Get(ctx);
```

Comment:
**Note the check for `NaN` values only applies to floating-point types and when `nan_is_null` is `true`, other types/cases can use the logic as it was.**

In the case of `ArrayData` there are 3 scenarios:
a) arr.GetNullCount() == arr.length  // All data is null
b) arr.GetNullCount() == 0               // No data is null
c) arr.MayHaveNulls() == true         // Some data is null

IIUC, the null bitmap of the input `ArrayData` is not guaranteed to be consistent with the data ([an `ArrayData` can be malformed](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/data.h#L226-L230) bc buffer values can be modified directly). Scenarios *(a)-(b)* invoke `arr.GetNullCount()` which iterates through all the `arr` values and updates the null bitmap accordingly.

Given that scenarios *(b)-(c)* are the common case and the array data has to be traversed to identify the `NaN` values, (as an optimization) I suggest to not check the null count at all. Nevertheless, only check for `NaN`s in non-null indices.

Reply:
I have pushed f404910 which contains the ArrayData validity for nulls (not sure if it's the best way to do). 
There is a way to test those changes in R ? As the ticket is also related to R language.

Reply:
No need to test kernel implementation in R (or Python), first bindings have to be put in place with the desired default option. The C++ kernels is invoked via the language bindings.

Reply:
@Christian8491 when this PR is ready, the C++ code is all reviewed and approved, and the CI is all green, please let me know and I can push a commit that changes the R bindings to use this new option.

Reply:
@pitrou The `is_null` compute function most of the time will iterate through the bitmap of the input `ArrayData` and when `nan_is_null` option is set, it can increase the null count. Would it be a good idea to update the null count of the input array (`arr.SetNullCount(...)`) as a side-effect of invoking this compute function?

Reply:
Why would you update the null count of the input array? I'm not sure I understand what you're proposing.

Reply:
I see, your point now, thanks! Modifying the input array does not makes sense because `nan_is_null` property applies only to the intermediate/output of the compute function. Please ignore these comments.